enhancements/api-review/gcp_user_defined_labels.md

```diff
@@ -0,0 +1,566 @@
+- GCP client libraries support two ways of providing tag details
+  1. NamespacedName: Tag value could be fed to APIs in the format
+     `<ORGANIZATION_ID>/<TAG_KEY_SHORT_NAME>/<TAG_VALUE_SHORT_NAME>` where
+      - ORGANIZATION_ID: Organization ID which is hosting the project under which OCP cluster
+     will be created and tag key, value resources exist.
+      - TAG_KEY_SHORT_NAME: Tag key short name is the display name provided during tag key
+     resource creation.
+      - TAG_VALUE_SHORT_NAME: Tag value short name is the display name provided during tag value
+     resource creation.
+  2. Name: Tag value could also be fed to the APIs in the format `tagValues/<TAG_VALUE_ID>` where
+      - TAG_VALUE_ID: Tag value ID is the generated numeric ID for the tag value resource using which
+     the tag key it is linked to can be identified.
+
+  GCP terraform API `google_tags_location_tag_binding` expects the tag value's `Name` whereas the GO API
+  `CreateTagBinding` accepts both tag value's `Name` and `NamespacedName`. The current approach for user
+  to input `ORGANIZATION_ID, TAG_KEY_SHORT_NAME, TAG_VALUE_SHORT_NAME` which has more readability, will require
+  1. For Terraform API([API doc](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/tags_tag_binding))
+      - To fetch the tag key's `Name` using the `short_name`.
+      - To fetch the tag value's `Name` using the `short_name` and the tag key's `Name`.
+      - Use the tags value's `Name` to create tag binding to a resource using `google_tags_location_tag_binding` API.
+  2. For GO API([API doc](https://github.com/googleapis/google-cloud-go/blob/resourcemanager/v1.8.2/resourcemanager/apiv3/tag_bindings_client.go#L174))
+      - To derive the tag value's `NamespacedName` using the aforementioned user provided parameters.
+      - Use the derived tag value's `NamespacedName` to create tag binding to a resource using `CreateTagBinding` API.
+
+  Another approach is for user to configure just the list of tag value's names of the form `tagValues/<TAG_VALUE_ID>`,
+  and it can be used directly with both terraform and GO APIs, and no additional processing on tags would be required.
+  But the tag values configured are in obscured form and it will be diffcult for mapping the tags configured and the
+  tags present on resources.
+
+  Would it be better to ask user to configure just the tag value name, and not Organization ID, tag key short name, though
+  it reduces the readability?
+
+### Test Plan
+- Upgrade/downgrade testing
+- Sufficient time for feedback
+- Available by default
+- Stress testing for scaling and label/tag update scenarios
+
+### Graduation Criteria
+#### Dev Preview -> Tech Preview
+- Feature available for end-to-end usage.
+- Complete end user documentation.
+- UTs and e2e tests are present.
+- Gather feedback from the users.
+
+#### Tech Preview -> GA
+N/A. This feature is for Tech Preview, until decided for GA.
```

Comment:
What featuregate will be used to gate this? Are we adding a new FeatureGate under the TechPreviewNoUpgrade FeatureSet? What testing will be in place before this is promoted to GA?

Reply:
We will add a TechPreviewNoUpgrade validation in installer [here](https://github.com/openshift/installer/blob/master/pkg/types/validation/installconfig.go#L1031) to allow configuration of tags and labels only when TP feature is set.

Full fledged testing will be carried out for TP implementation, any user suggestions/feedback incorporated will be tested for GA along with the regression tests. Please recommend if other tests are required or need to be looked at.

Reply:
Features should be blocked by `FeatureGates` not `FeatureSets`, otherwise a use cannot turn on/customise their selection to individual features. Also, the controllers in the cluster need to be able to understand if the feature is enabled or not. We don't want them to run the tag syncing code if it's not enabled.

So you should be adding a [FeatureGate](https://github.com/openshift/api/blob/master/config/v1/feature_gates.go) and blocking the functionality on that rather than on `TechPreviewNoUpgrade`

Reply:
Since tags and labels can be configured only during cluster creation, through `FeatureSets`, I was thinking if `TechPreviewNoUpgrade` is not set, tags/labels configuration will not be allowed and cluster creation will fail with error.

And for operators since Infrastructure would not have tags & labels it will be skipped.
But a provision for configuring tags in machine provider operator like for AWS, Azure tags would have to be controlled through `FeatureGates` as it's not supported currently, though operator supports configuring labels.

Reply:
By relying on the `FeatureSet`, you cannot allow users to customise their install. What if, for example, the user wanted to try this feature without other `TechPreviewNoUpgrade` feature gates? Normally they would do that using `CustomNoUpgrade` and selecting only the features they want, with this approach, that won't work

Reply:
Yeah, I understand, I will make changes to use `FeatureGates`